### PR TITLE
New Web UI

### PR DESCRIPTION
**Pull request by tejasbubane**
This is the implementation of the new UI being discussed here: https://github.com/whitesmith/rubycritic/issues/173
I also took the liberty to upgrade some of the javascript dependencies.

This is a combined effort from @cybrilla and we are loving it ❤️
Run it and let us know what you think.!

Closes: https://github.com/whitesmith/rubycritic/issues/173

### CODE REVIEW

File: lib/rubycritic/core/analysed_module.rb [hunk 1]
```diff
@@ -19,6 +23,22 @@ def path
       @path ||= pathname.to_s
     end
 
+    def file_location
+      file_path[0...-1].join('/')
```
Comment (troessner) [inline lib/rubycritic/core/analysed_module.rb]: Did we actually ever discuss if we support Windows? If yes, this should be `File::SEPARATOR`.

**Comment by troessner** [comment]
@lucasmazza @Onumis are we actually supporting windows? That's something we should clarify and put in our README.
The Ruby code looks good for me to merge, the JS stuff looks ok for me as well with one exception (pls see my comment).

@tejasbubane amazing job, thanks a lot for this!

File: lib/rubycritic/generators/html/assets/javascripts/application.js [hunk 2]
```diff
@@ -1,107 +1,240 @@
...
+        dataLabels: {
+          enabled: false
+        }
+      }]
+    },
+    function(chart) { // on complete
+      var textX = chart.plotLeft + (chart.plotWidth  * 0.5);
+      var textY = chart.plotTop  + (chart.plotHeight * 0.5);
+      var span = '<span id="pieChartInfoText" style="position:absolute; text-align:center;">';
+      span += '<span style="font-size: 32px">'+((score/100)*4).toFixed(2)+'</span><br>';
```
Comment (Onumis) [inline lib/rubycritic/generators/html/assets/javascripts/application.js]: Although smaller numbers look better, I think we should still rely on our _0 to 100_ score system (https://github.com/whitesmith/rubycritic/blob/master/docs/core-metrics.md#score)

File: lib/rubycritic/generators/html/assets/javascripts/application.js [hunk 3]
```diff
@@ -1,107 +1,240 @@
...
+      span += '<span style="font-size: 16px">GPA</span>';
```
Comment (Onumis) [inline lib/rubycritic/generators/html/assets/javascripts/application.js]: And instead of `GPA` we can display `/100` or `of 100`.

File: lib/rubycritic/generators/html/templates/code_index.html.erb [hunk 4]
```diff
@@ -1,30 +1,45 @@
...
+    <div class="Page_Title">
+      <h3>Code</h3>
+    </div>
+    <!--End Page Title -->
+    <div class="Content_Wrapper">
+      <table id="codeTable" class="table code sortable-table">
+        <thead>
+          <tr>
+            <% unless Config.suppress_ratings %>
+              <th width="10%" class="table-header">Rating<span class="sort-type"></span></th>
```
Comment (Onumis) [inline lib/rubycritic/generators/html/templates/code_index.html.erb]: I noticed the default order in this table changed from "name" to "rating". And I actually like it, but I think it would be best if the order was reversed, so that the lowest ratings are at the top.

File: lib/rubycritic/generators/html/templates/code_file.html.erb [hunk 5]
```diff
@@ -1,38 +1,65 @@
...
+              <div><span class="metric"><%= @analysed_module.churn %></span><small> loc/method</small></div>
+            </div>
+            <div class="col-md-3">
+              <div><span class="metric"><%= @analysed_module.complexity %></span><small> complexity</small></div>
+              <div><span class="metric"><%= @analysed_module.duplication %></span><small> duplications</small></div>
+            </div>
+            <div class="col-md-3">
+              <small><%= @analysed_module.complexity %> complexity</small> <br/ >
+              <small><%= @analysed_module.duplication %> duplications</small>
+            </div>
```
Comment (Onumis) [inline lib/rubycritic/generators/html/templates/code_file.html.erb]: I think this `div` is duplicated.

File: lib/rubycritic/core/analysed_module.rb [hunk 6]
```diff
@@ -19,6 +23,22 @@ def path
...
+    def file_name
+      file_path[-1]
+    end
+
+    def file_path
+      path.split('/')
+    end
+
+    def line_count
+      `wc -l "#{path}"`.strip.split(' ')[0].to_i
```
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: This is introducing extra behaviour that may not work on other platforms, I believe.

Maybe this could be moved to another PR that attempts to solve this in a more cross-platform way?

File: lib/rubycritic/core/analysed_module.rb (hunk 1 above)
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: It should support Windows (at least I think it used to).

And `File::SEPARATOR` is unnecessary as an `analysed_module` has a `pathname` attribute which is of the type, uh, [Pathname](http://ruby-doc.org/stdlib-1.9.3/libdoc/pathname/rdoc/Pathname.html) 😄 

This method should simple be

```
pathname.dirname
```

File: lib/rubycritic/core/analysed_module.rb [hunk 7]
```diff
@@ -19,6 +23,22 @@ def path
       @path ||= pathname.to_s
     end
 
+    def file_location
+      file_path[0...-1].join('/')
+    end
+
+    def file_name
+      file_path[-1]
```
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: `pathname.basename` or `pathname.basename.sub_ext('').to_s` if you don't want the file extension, as seen in the `Location` class.

File: lib/rubycritic/core/analysed_module.rb [hunk 8]
```diff
@@ -19,6 +23,22 @@ def path
...
+    end
+
+    def file_path
```
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: I believe this will now be an unnecessary method 😄

File: lib/rubycritic/generators/html/templates/code_index.html.erb [hunk 9]
```diff
@@ -1,30 +1,45 @@
...
+                <ul class="nav nav-pills">
+                  <li role="presentation" >
+                    <a href="<%= file_path(analysed_module.pathname.sub_ext('.html')) %>"><%= analysed_module.name %></a>
+                  </li>
+                </ul>
+              </td>
+              <td><%= analysed_module.churn %></td>
+              <td><%= analysed_module.complexity %></td>
+              <td><%= analysed_module.duplication %></td>
+              <td><%= analysed_module.smells.length %></td>
```
Comment (guilhermesimoes) [inline lib/rubycritic/generators/html/templates/code_index.html.erb]: We can now use the new `smells_count` method.

**Comment by tejasbubane** [comment]
Review comments fixed, please check @Onumis

**Comment by Onumis** [comment]
Alright! Looks good to me!
Please, [squash changes into a single commit](https://github.com/whitesmith/rubycritic/blob/master/CONTRIBUTING.md#writing-some-code) :)

**Comment by rohitcy** [comment]
@Onumis Squashed all the commits, we have also fixed the logo text 'Ruby Critic' to 'RUBYCRITIC'(No Spacing) it looks like this now:

screen shot 2016-11-23 at 8 33 00 pm

**Comment by Onumis** [comment]
Thanks! Gonna release it now! :smile:

**Comment by tejasbubane** [comment]
@Onumis Great! 🎉  Sending a PR to update the docs with new images.

File: lib/rubycritic/core/analysed_module.rb (hunk 6 above)
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: https://gist.github.com/guilhermesimoes/d69e547884e556c3dc95
